dataplicity/clockcheck.py

```diff
@@ -0,0 +1,44 @@
+"""
+On RPi, threading.Event objects block indefinitely when the clock
+changes, which causes agent to hang. This thread kills the current
+process when the clock changes. Supervisor will then restart agent,
+and we will recover.
+
+It's not nice at all. Probably the result of an OS bug. Hopefully,
+there will be a better solution in future.
+
+"""
+
+import os
+import signal
+from time import sleep, time
+from threading import Thread
+
+
+CHECK_PERIOD = 1.0
+MAX_CLOCK_DISCREPENCY = 10.0
+
+
+class ClockCheckThread(Thread):
+    """Thread to restart agent when the clock changes."""
+
+    def __init__(self):
+        super(ClockCheckThread, self).__init__()
+        self.daemon = True
+        self.running = True
+
+    def run(self):
+        """Calls `on_fail` when the system clock changes."""
+        while self.running:
```

Comment:
Normally a threading event would be preferable to a boolean here, but we're working around Event objects being fubar when the clock changes.